<!DOCTYPE html>
<html lang="en">

<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  <script src="bootstrap.min.js"></script>
  <link rel="stylesheet" type="text/css" href="bootstrap.min.css" />
  <link rel="shortcut icon" type="image/x-icon" href="LogoHK220X150.png" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <link rel="stylesheet" type="text/css" href="mystyle.css" />

  <title>About HRT Atlas v1.0</title>
  <meta name="Description"
    content="HRT Atlas v1.0 is a web-based tool for searching human and mouse housekeeping transcripts and genes." />

  <meta itemprop="name" content="Housekeeping genes database" />
  <meta itemprop="description" content="Human and mouse housekeeping genes for qPCR normalization.
    Users can search for the most suitable tissue specific reference genes (endogenous controls) and efficient primers to normalize their qPCR experiments.
    Diseases and small molecules that can modify expression stability." />

  <meta itemprop="image" content="pipeline2.png" />
  <link rel="publisher" href="https://plus.google.com/+www.housekeeping.unicamp.br" />
  <meta property="og:url" content="www.housekeeping.unicamp.br" />
  <meta property="og:site_name" content="Housekeeping genes database" />
  <meta property="og:description" content="Human and mouse housekeeping genes for qPCR normalization.
    Users can search for the most suitable tissue specific reference genes (endogenous controls) and efficient primers 
    to normalize their qPCR experiments. Diseases and small molecules that can modify expression stability." />

  <meta property="og:type" content="website" />
  <meta property="og:image" content="pipeline2.png" />
  <meta name="twitter:url" content="www.housekeeping.unicamp.br" />
  <meta name="twitter:title" content="Housekeeping genes database" />
  <meta name="twitter:description" content="Human and mouse housekeeping genes for qPCR normalization.
    Users can search for the most suitable tissue specific reference genes (endogenous controls) and efficient primers to 
    normalize their qPCR experiments. 
    Diseases and small molecules that can modify expression stability." />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:image" content="pipeline2.png" />
  <meta name="twitter:creator" content="@BidossessiH" />
  <meta name="author" content="Bidossessi Wilfried Hounkpe" />
  <meta name="robots" content="index" />
  <meta name="google-site-verification" content="DGWLR3MQZ2V6rHusJy1Hb_VWOqUcVCFqv1aNGNOvmiQ" />
  <script src="https://www.googletagmanager.com/gtag/js?id=UA-137785214-1">async</script>

  <!--Global site tag (gtag.js) - Google Analytics -->
  <script>

    window.dataLayer = window.dataLayer || [];
    function gtag() { dataLayer.push(arguments); }
    gtag('js', new Date());
    gtag('config', 'UA-137785214-1');
  </script>

</head>


<body>



  <nav class="navbar navbar-expand-md navbarStyle navbar-light" style="width:100%;">
    <a class="navbar-brand logoviz" href="http://www.housekeeping.unicamp.br/?homePageGlobal">
      <img src="LogoHKNav2.png" alt="Logo" class="d-inline-block align-center logowidth"> <span
        style="text-align:center; font-weight: bold">Housekeeping and Reference Transcript Atlas</span>
    </a>

    <button class="navbar-toggler hidden-md-up float-md-right" type="button" data-toggle="collapse"
      data-target="#collapsibleNavbar">
      <span class="navbar-toggler-icon"></span>
    </button>
    <div class="collapse navbar-collapse navbar-toggleable-sm navbarcustom" id="collapsibleNavbar">
      <ul class="nav navbar-nav">
        <li class="nav-item">
          <a class="nav-link" href="http://www.housekeeping.unicamp.br/?homePageGlobal">Home</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="http://www.housekeeping.unicamp.br/?about">About</a>
        </li>

        <li class="nav-item">
          <a class="nav-link" href="http://www.housekeeping.unicamp.br/?download"
            onclick="ga('send', 'event', 'Menu', 'access', 'Download');">Downloads</a>
        </li>

        <li class="nav-item dropdown">
          <a class="nav-link dropdown-toggle" href="#" id="navbarDropdownMenuLink" data-toggle="dropdown"
            aria-haspopup="true" aria-expanded="false">
            Gene expression
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
            <a class="dropdown-item" href="http://www.housekeeping.unicamp.br/?visualization" style="text-align:justify"
              onclick="ga('send', 'event', 'Menu', 'access', 'Visualize Human');">Human housekeeping genes expression
              <br>
              in different tissues</a>
            <a class="dropdown-item" href="http://www.housekeeping.unicamp.br/?visualizationMouse"
              style="text-align:justify" onclick="ga('send', 'event', 'Menu', 'access', 'Visualize Mouse');">Mouse
              housekeeping genes expression
              <br>
              in different tissues</a>

          </div>
        </li>

        <li class="nav-item">
          <a class="nav-link" href="http://www.housekeeping.unicamp.br/?modifiers"
            onclick="ga('send', 'event', 'Menu', 'access', 'Modifiers');">Expression
            Modifiers</a>
        </li>

      </ul>
    </div>
  </nav>

  <!--Show dropdown menu on hover-->
  <script>
    $('ul.nav li.dropdown').hover(function () {
      $(this).find('.dropdown-menu').stop(true, true).delay(200).fadeIn(500);
    }, function () {
      $(this).find('.dropdown-menu').stop(true, true).delay(200).fadeOut(500);
    });

  </script>

  <div class="container">
    <br>
    <h1>Housekeeping genes and reference genes/transcripts</h1>
    <br>
    <p class="text-justify">Despite the importance of housekeeping genes in the calibration of gene expression
      (reference genes), as well as the understanding of many genomic and evolutionary features, important discrepancies
      have been observed in studies that previously identified these genes. HRT Atlas addresses some of the previously
      observed limitations in the identification of these genes, and offers a more accurate database of human and
      mouse housekeeping genes and tissue selective reference genes/transcripts.
    </p>
    <br>
    <img src="pipeline2.png" style="width: 100%;" alt="Housekeeping genes and reference genes/transcripts">

    <h1>Overview of Housekeeping and Reference Transcript Atlas pipeline</h1>


    <br>
    <br>


    <div id="accordion">
      <div class="card">
        <div class="card-header">
          <a class="card-link" data-toggle="collapse" href="#collapseintro">
            <h1>Housekeeping and Reference Transcript Atlas (HRT Atlas v1.0)</h1>
          </a>
        </div>
        <div id="collapseintro" class="collapse show" data-parent="#accordion">
          <div class="card-body" style="text-align:justify; line-height: 1.6;">
            <p>
              Housekeeping and Reference Transcript Atlas is a web-based tool with user-friendly interface for searching
              human and
              mouse housekeeping transcripts and genes. The workflow for the generation of the human database included
              the analysis of more than 11,000 high quality RNA-seq datasets, obtained from <a
                href="https://gtexportal.org/home/" target="_blank">GTEx</a> (version 7),
              a comprehensive public resource of tissue-specific gene expression. Data from 52 different tissue and
              cell types were obtained. In addition, data from 50 tissues and cell types were obtained from <a
                href="https://amp.pharm.mssm.edu/archs4/index.html" target="_blank">ARCHS4</a>,
              a mining tool for public RNA-seq experiments.
              The generation of mouse database included the analysis of 507 high quality RNA-seq datasets manually
              curated from GEO database. Raw read counts were obtained from ARCHS4 repository. Samples of 14 tissues
              and cells types from wild type healthy control mice were included in the workflow.
            </p>
          </div>
        </div>
      </div>

      <div class="card">
        <div class="card-header">
          <a class="card-link" data-toggle="collapse" href="#collapse1">
            <h1>What does Housekeeping and Reference Transcript Atlas offer?</h1>
          </a>
        </div>
        <div id="collapse1" class="collapse" data-parent="#accordion">
          <div class="card-body" style="text-align:justify; line-height: 1.6;">
            <p>
              Users can download a complete list of housekeeping transcripts and genes built from our analysis of
              more than 11,000 high quality human RNA-seq datasets and 507 mouse datasets. Alternatively, they can
              search for the most suitable candidate reference transcripts to normalize their qPCR experiments. These
              candidate reference
              transcripts represent the most stable transcripts for each of the different tissues or cell types
              available in our database (102 and 22 for human and mouse respectively). In addition, specific primers
              for some of these transcripts are also provided and validated following MIQE guidelines. In the download
              page, it also provides a short
              list of reference transcripts stably expressed at low variability in all of the tissue or cell types.
              This list is provided to normalize tissue or cell types that cannot be accessed from our database.
              As diseases and small molecules are known to affect gene expression variability,
              we also integrated and manually curated three functional associations datasets from <a
                href="http://amp.pharm.mssm.edu/Harmonizome/" target="_blank">Harmonizome</a>
              database. These datasets show for each human housekeeping gene, a list of known diseases and
              molecules that can modify their expression stability (based on differential expression) and intend
              to guide users in the choice of suitable reference genes for qPCR normalization.

              In addition, HRT Atlas has been integrated with EpiRegio REST API.
              <a href="https://epiregio.de/">EpiRegio</a> is a resource of Regulatory
              EleMents (REMs) that are able to regulate the expression of their
              target genes in different cell types and tissues.
            </p>
          </div>
        </div>
      </div>

      <div class="card">
        <div class="card-header">
          <a class="collapsed card-link" data-toggle="collapse" href="#collapse2">
            <h1>Gene annotation in Ensembl</h1>
          </a>
        </div>
        <div id="collapse2" class="collapse" data-parent="#accordion">
          <div class="card-body" style="text-align:justify; line-height: 1.6;">
            <p>
              In Housekeeping and Reference Transcript Atlas transcripts are identified by Ensembl ID (Ensembl
              annotation).
              <a href="https://www.ensembl.org/index.html" target="_blank">Ensembl</a> is an open access project that
              integrates gene annotation, comparative genomics,
              variant annotation and annotation of regulatory features of vertebrate genomes.
              A gene is a genomic locus where transcription occurs and which may have one or more protein-coding
              isoform and/or non-coding transcripts. In Ensembl annotation process, a gene includes any spliced
              transcripts with overlapping coding sequence. That way, transcripts clustered into the same gene differ
              in their genomic coordinates (transcription start and end sites), splice events and exons number and
              sequences. The image below shows an illustrative example of transcripts that belong to GAPDH. Non-coding
              transcripts are shown in blue. All other are transcripts are annotated protein-coding transcripts.
            </p>
            <br>
            <br>
            <img src="gapdh.png" style="width:100%">
            <br>
            <br>
            <h3>Transcript models of <a
                href="http://Apr2019.archive.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000111640;r=12:6534512-6538374"
                target="_blank">

                <i>GAPDH</i>
              </a>
              </h2>
          </div>
        </div>
      </div>
      <div class="card">
        <div class="card-header">
          <a class="collapsed card-link" data-toggle="collapse" href="#collapse3">
            <h1>Criteria for housekeeping genes and transcripts selection</h1>
          </a>
        </div>
        <div id="collapse3" class="collapse" data-parent="#accordion">
          <div class="card-body" style="text-align:justify; line-height: 1.6;">
            <p>
              Housekeeping genes have been classically defined as genes that are required for the maintenance
              of essential functions of any cell type. Hence, they are expected to be constitutively expressed
              in all cells and at any development stage. In Housekeeping and Reference Transcript Atlas, a housekeeping
              gene
              is defined as a gene that has at least one of its protein-coding transcripts expressed in all cell
              types or tissues, with low variability (and non-zero level). Each of these transcripts may be specific
              to a subset of tissue or cell types. Transcripts belonging to these housekeeping genes are considered
              housekeeping transcripts. The housekeeping transcripts must have additionally a non-zero expression
              in all of the samples analyzed.
            </p>
            <br>

            The following criteria were applied to select the list of housekeeping genes, which should fulfill all three
            criteria:
            <br>
            <ul>
              <li>The gene must be expressed at non-zero level in all tissue and cell types included in our analysis (52
                and 14 cells and tissues for human and mouse respectively). This means that at least one of its
                transcripts must have an expression level higher than 1 <a
                  href="https://www.ebi.ac.uk/training/online/glossary/rpkm" target="_blank">RPKM</a> (reads per
                kilobase
                million)</li>
              <li>Variability of gene expression should be low within all tissues and cell types, as evidence by a
                standard deviation of the log2 RPKM &#8826; 1</li>
              <li>Maximum fold change (MFC), represented by the ratio between log2 of maximum and log2 of mean RPKM of
                any
                of its transcripts, must be lower than 2. Thus, only genes whose expression level in all samples fall
                within a
                very narrow range are classified as housekeeping genes.</li>

            </ul>


            Only transcripts with well-supported transcript models were included in the database. As pseudogenes
            expression may affect the variability of their associated genes, genes with known pseudogenes were also
            excluded from the lists of reference transcripts proposed for qPCR experiment normalization.

          </div>
        </div>
      </div>

      <div class="card">
        <div class="card-header">
          <a class="collapsed card-link" data-toggle="collapse" href="#collapse4">
            <h1>How results are ranked?</h1>
          </a>
        </div>
        <div id="collapse4" class="collapse" data-parent="#accordion">
          <div class="card-body" style="text-align:justify; line-height: 1.6;">
            <p>
              We proposed the term <a class="font-weight-bold">reference transcript</a> to define transcripts that are
              suitable for qPCR normalization
              in a particular tissue type. These transcripts were selected from the list of the newly described HK
              transcripts and have
              also a very low variability (log2 of maximum RPKM &#8826; 2 x log2 of minimum RPKM and standard deviation
              &#8826; 0.5) in specific
              tissue types, but not necessary in all of the tissues included in the database.
              Candidate reference transcripts stability was ranked based on mean RPKM, standard deviation of log2 RPKM
              and
              MFC value.
              Giving n transcripts, three intermediate scores (k=3) with equal weight are generated (score-1, score-2,
              and
              score-3 for RKPM, standard deviation and MFC, respectively for each transcript). Score-1 assigns n to the
              highest
              expressed transcript, estimated by mean RPKM, and 1 to the lowest expressed. Similarly, for score-2 and
              score-3 the
              highest scoring transcripts are the one that have the lowest standard deviation or the lowest MFC. Given
              sct,i
              the score of the transcript t for the i-th rank, we express the Score Product (SP) via the geometric mean:
            </p>
            <br>
            <br>
            <img src="formula.png">
            <br>
            <br>
            The highest scoring transcript is one that has the highest SP and is assigned to the first position (rank =
            1).
            That way, the n-th transcript is the one that has the lowest SP. The Score Product is proportional to the
            variability.


          </div>
        </div>
      </div>

      <div class="card">
        <div class="card-header">
          <a class="collapsed card-link" data-toggle="collapse" href="#collapse5">
            <h1>Housekeeping and Reference Transcript Atlas pipeline</h1>
          </a>
        </div>
        <div id="collapse5" class="collapse" data-parent="#accordion">
          <div class="card-body" style="text-align:justify; line-height: 1.6;">
            <p>
              Data included in Housekeeping and Reference Transcript Atlas pipeline are provided in useful format and
              generated by GTEx
              and ARCHS4 workflows. Transcript read counts of healthy human tissues and cells types were extracted fom
              GTEx V7. For mouse data analysis, GEO series extracted from ARCHS4 HD5F format were manually curated to
              only
              include samples from wild type C57BL/6 healthy mice. Read counts were then extracted from ARCHS4
              repository
              after the following additional restrictive criteria:
            </p>
            <ul>
              <li>Sequencing depth must be equal or higher than 20,000,000 reads</li>
              <li>Alignment rate provided by ARCHS4 must be higher than 70%</li>
              <li>Library layout must be generated from mRNA enrichment protocol</li>
              <li>Paired end read library must be constructed. This enables an accurate detection of transcript
                isoforms.
              </li>
            </ul>
            <br>
            <p>
              Read counts are then library-size normalized, and RPKM calculated using edgeR package in R environment.
              Lists of housekeeping genes and transcripts were generated based on the criteria described above
              (<i>“Criteria for housekeeping genes and transcripts selection”</i> section).
              The housekeeping transcript list was then used to provide a list of suitable references for qPCR
              experiment
              normalization for each tissue and cell type. Some transcripts specific primers were also designed. All of
              the generated matrixes were used to construct a SQLite database.
              Housekeeping and Reference Transcript Atlas is a web tool developed with Shiny package and hosted by Shiny
              Server. The application front-end is implemented by HTML, CSS and Bootstrap 4. All of the components were
              encapsulated in a deployable Docker container running on an Unix-based operating system with Apache HTTP
              server. Users
              instructions communicated to the container are processed by R and the request is sent to the SQLite
              database
              mounted as a container volume. Housekeeping and Reference Transcript Atlas offers a user friendly and
              reactive interface to provide the best using experience.
            </p>
          </div>
        </div>
      </div>
    </div>

    <div class="card">
      <div class="card-header">
        <a class="collapsed card-link" data-toggle="collapse" href="#collapse6">
          <h1>Housekeeping and Reference Transcript Atlas interactive website</h1>
        </a>
      </div>
      <div id="collapse6" class="collapse" data-parent="#accordion">
        <div class="card-body" style="text-align:justify; line-height: 1.6;">
          <img src="screenshotHome.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot of Housekeeping and Reference Transcript Atlas home page</h3>
          <br>
          <br>

          <img src="screenshotHomehs.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot of search interface of human database</h3>

          <br>
          <br>

          The search interface allows searching suitable reference transcripts for qPCR normalization by selecting
          one of the 82 tissues or cell types included in our database. By clicking on the search box, the full list
          of cells and tissues is automatically displayed on the screen ("Screenshot showing inputs of the search box").
          <br>
          From this interface users can also access and download the complete list of housekeeping transcripts and
          housekeeping genes as shown below. A similar interface is provided for mouse database.
          <br>
          <br>

          <img src="screenshotHomesearch.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot showing inputs of the search box</h3>

          <br>
          <br>

          The search box provides an autocomplete option which makes easy the selection of tissue or cell
          type included in the database. This list is specific to the selected specie.

          <br>
          <br>

          <img src="screenshotHomeresults.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot of the results page</h3>

          <br>
          <br>
          <p>
            The results page shows the list of reference transcripts for the selected tissue or
            cell type. Each transcript is identified by its Ensembl code, followed by the gene symbol.
            The expression level of each transcript is expressed as the mean value of reads per kilobase million
            (RPKM) for the specific tissue or cell type. Variability and maximum fold-change are expressed as the
            standard deviation of the log2 RPKM, and the ratio of log2(maximum RPKM)/ log2(mean RPKM), respectively.
            By default, results are ranked by the criteria described above (“How results are ranked?” section).
            Alternatively, transcripts can be ranked
            by the user by clicking in the arrow next to any of these parameters.
            Users can select the number of transcripts to be shown per page, and then download the full list displayed
            on the screen in their preferred format (csv, Excel or Pdf). Finally user is allowed to
            filter de candidate reference transcript table by disabling the logarithmic normalization option.
            A minimum valor of RPKM can also be defined. The default is 30.
          </p>
          <br>
          <br>

          <img src="screenshotHomeprimer.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot showing specific primers table</h3>

          <br>
          <br>

          <img src="screenshotHomevalidation.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot showing specific primers qPCR validation</h3>
          <br>
          <br>
          <img src="screenshotHomeEpiregio.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot showing regulatory elements retrieved fom EpiRegio server.</h3>

          <br>
          <br>

          <img src="screenshotHarmonizome.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot showing example of functional associations datasets extracted from Harmonizome database</h3>
          From this section, users can access the following functional associations datasets extracted from Harmonizome
          database:
          <ul>
            <li>
              <a href="http://amp.pharm.mssm.edu/Harmonizome/dataset/GEO+Signatures+of+Differentially+Expressed+Genes+for+Diseases"
                target="_blank">
                GEO Signatures of Differentially Expressed Genes for Diseases
              </a>

            </li>

            This dataset shows for a selected gene, a list of diseases that can induce its differential expression.
            <li>
              <a href="http://amp.pharm.mssm.edu/Harmonizome/dataset/GEO+Signatures+of+Differentially+Expressed+Genes+for+Small+Molecules"
                target="_blank">
                GEO Signatures of Differentially Expressed Genes for Small Molecules
              </a>
            </li>
            It shows small molecule perturbations that can change the expression of a particular gene. Users must avoid
            using a gene for qPCR normalization purpose if their molecules of interest are listed
            in this dataset.

            <li>
              <a href="http://amp.pharm.mssm.edu/Harmonizome/dataset/CMAP+Signatures+of+Differentially+Expressed+Genes+for+Small+Molecules"
                target="_blank">
                Connectivity Map Signatures of Differentially Expressed Genes for Small Molecules
              </a>
            </li>
            It shows small molecules known to induce the differential expression of the selected gene.
            This gene must also be avoided for normalization of qPCR if the molecule of interest is listed.
          </ul>
          <a href="http://amp.pharm.mssm.edu/Harmonizome/about" target="_blank">Learn more about these datasets</a>



          <br>
          <br>

          <img src="screenshotVisualization.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot showing "Visualization" section</h3>
          Users can access the Visualization page from the navigation bar. When available, they can access the
          description
          of
          a selected housekeeping gene, its official name and its synonymes retrieved from the NCBI's
          portal of gene-specific content based on RefSeq project. Gene expression across tissue and cell types is also
          provided.



          <br>
          <br>

          <img src="screenshotHomedownload.png" style="width:100%; box-shadow: 0px 2px 16px #999;">
          <br>
          <br>
          <h3>Screenshot of download</h3>

          <br>
          <br>
          Download page provides access to the complete list housekeeping genes and transcripts, and a list of most
          stable
          transcripts. Data may be saved in csv or RData format.

          <br>
          <br>

          <img src="screenshotHomesmall.png" style="box-shadow: 0px 2px 16px #999;" class="screenshotSmall">
          <br>
          <br>
          <h3>Screenshot from small device</h3>

          <br>
          <br>
          Housekeeping and Reference Transcript Atlas is fully responsive in small and medium screen width. User can
          access all
          functionalities of the database from any size of screen.
        </div>
      </div>
    </div>
  </div>

  </div>
  </div>

  </div>
  <br>
  <br>

  <!--Footer-->
  <div id="footer" class="shadow-top">
    <br />
    <div class="col-xs-3 col-lg-2 footerdiv1">
      <a href='https://www.unicamp.br' target='_blank'><img src='Unicamp.png' class='imageFooter'></a>
      <a href='https://www.hemocentro.unicamp.br' target='_blank'><img src='Hemocentro.png'
          class='imageFooter imageFooter1'></a>
    </div>
    <div class="col-xs-4 col-lg-5 bordst footerdiv">
      <div sytle="margin:auto" class="footerdiv2">
        Affiliation
        <br />
        University of Campinas
        <br />
        Laboratory of Hemostasis and Inflammation - LAHI
        <br />
        Hemocentro Campinas/Brazil
      </div>
    </div>
    <div class="col-xs-4 col-lg-5 bordst footerdiv">
      <div sytle="margin:auto" class="footerdiv2">
        Contact
        <br />
        For questions, problems or suggestions, please send a mail to Bidossessi Wilfried: openbiotools@gmail.com
      </div>
      <br />
      <a href='https://github.com/Bidossessih/HRT_Atlas' target='_blank'><img src='github.png'
          style='width: 25px; float: right;'></a>
    </div>

  </div>

  <script>
    $(document).ready(function () {
      $('[data-toggle="tooltip"]').tooltip();
    });
  </script>

  <script>
    $('.collapse').collapse({ toggle: false })
  </script>

</body>

</html>